pkg/nsx/services/vpc/vpc.go

```diff
@@ -519,42 +522,59 @@ func (s *VPCService) DeleteLBPool(path string) error {
 	log.Info("successfully deleted NCP created lbPool", "lbPool", path)
 	return nil
 }
+
 func (s *VPCService) IsSharedVPCNamespaceByNS(ns string) (bool, error) {
-	shared_ns, err := s.getSharedVPCNamespaceFromNS(ns)
+	_, sharedNamespaceObj, err := s.resolveSharedVPCNamespace(ns)
 	if err != nil {
 		return false, err
 	}
-	if shared_ns == "" {
+	if sharedNamespaceObj == nil {
 		return false, nil
 	}
-	if shared_ns != ns {
+	if sharedNamespaceObj.Name != ns {
 		return true, nil
 	}
 	return false, err
 }
 
-func (s *VPCService) getSharedVPCNamespaceFromNS(ns string) (string, error) {
+func (s *VPCService) getNamespace(name string) (*v1.Namespace, error) {
 	obj := &v1.Namespace{}
-	if err := s.Client.Get(ctx, types.NamespacedName{
-		Name:      ns,
-		Namespace: ns,
-	}, obj); err != nil {
-		log.Error(err, "failed to fetch namespace", "Namespace", ns)
-		return "", err
+	if err := s.Client.Get(ctx, types.NamespacedName{Name: name}, obj); err != nil {
+		log.Error(err, "Failed to fetch namespace", "Namespace", name)
+		return nil, err
+	}
+	return obj, nil
+}
+
+// resolveSharedVPCNamespace will resolve the Namespace relationship based on VPC sharing,
+// whether a shared VPC Namespace exists.
+func (s *VPCService) resolveSharedVPCNamespace(ns string) (*v1.Namespace, *v1.Namespace, error) {
+	obj, err := s.getNamespace(ns)
+	if err != nil {
+		return nil, nil, err
 	}
 
 	annos := obj.Annotations
 	// If no annotaion on ns, then this is not a shared VPC ns
 	if len(annos) == 0 {
-		return "", nil
+		return obj, nil, nil
 	}
 
 	// If no annotation nsx.vmware.com/shared_vpc_namespace on ns, this is not a shared vpc
-	shared_ns, exist := annos[common.AnnotationSharedVPCNamespace]
+	sharedNs, exist := annos[common.AnnotationSharedVPCNamespace]
 	if !exist {
-		return "", nil
+		return obj, nil, nil
+	}
+	if sharedNs == ns {
+		return obj, obj, nil
+	}
+	sharedNamespace, err := s.getNamespace(sharedNs)
+	if err != nil {
+		// if sharedNamespace does not exist, add this case for security,
+		// It shouldn't happen that a shared Namespace doesn't exist but is used as an annotation on another Namespace.
+		return nil, nil, err
 	}
-	return shared_ns, nil
+	return obj, sharedNamespace, nil
```

Comment:
Likewise, it should be
`return nil, sharedNamespace, nil`
because this case is to return ns for shared VPCs not for non-shared VPCs